tools/ts_library_builder/ast_util.ts

```diff
@@ -0,0 +1,238 @@
+import { readFileSync } from "fs";
+import {
+  ExportDeclaration,
+  ImportDeclaration,
+  InterfaceDeclaration,
+  JSDoc,
+  NamespaceDeclaration,
+  NamespaceDeclarationKind,
+  Project,
+  PropertySignature,
+  SourceFile,
+  StatementedNode,
+  TypeGuards,
+  VariableStatement,
+  VariableDeclarationKind
+} from "ts-simple-ast";
+
+// It is too complicated to deal with only what is used out of `msg_generated`
+// so we deal with it in a _hardcoded_ way.
+const MSG_GENERATED_SPECIFIER = "gen/msg_generated";
+export const INTERNAL_GENERATED_SPECIFIER = "@internal/generated";
+
+/**
+ * Add a property to an interface
+ */
+export function addInterfaceProperty(
+  interfaceDeclaration: InterfaceDeclaration,
+  name: string,
+  type: string,
+  jsdocs?: JSDoc[]
+): PropertySignature {
+  const interfaceProperty = interfaceDeclaration.addProperty({ name, type });
+  if (jsdocs) {
+    interfaceProperty.addJsDocs(jsdocs.map(jsdoc => jsdoc.getText()));
+  }
+  return interfaceProperty;
+}
+
+/**
+ * Add a namespace to a source file based on an original source file
+ */
+export function addNamespaceDeclaration(
+  namespace: string,
+  sourceFile: SourceFile,
+  destFile: SourceFile,
+  rootPath: string
+) {
+  const existingNamespace = destFile.getNamespace(`"${namespace}"`);
+  if (existingNamespace) {
+    return existingNamespace;
+  }
+
+  function rewriteDeclaration(
+    declaration: ExportDeclaration | ImportDeclaration
+  ): void {
+    const moduleSpecifier = declaration.getModuleSpecifier();
+    const moduleSourceFile = declaration.getModuleSpecifierSourceFile();
+    if (moduleSpecifier && moduleSourceFile) {
+      const moduleSpecifierText = moduleSpecifier.getLiteralValue();
+      if (moduleSpecifierText.startsWith("./")) {
+        const newModuleSpecifier = moduleSpecifierText.replace(
+          /^\.\//,
+          "@internal/"
+        );
+        addNamespaceDeclaration(
+          newModuleSpecifier,
+          moduleSourceFile,
+          destFile,
+          rootPath
+        );
+        declaration.setModuleSpecifier(newModuleSpecifier);
+      } else if (moduleSpecifierText === MSG_GENERATED_SPECIFIER) {
+        declaration.setModuleSpecifier(INTERNAL_GENERATED_SPECIFIER);
+      }
+    }
+  }
+
+  sourceFile.getExportDeclarations().forEach(rewriteDeclaration);
+  sourceFile.getImportDeclarations().forEach(rewriteDeclaration);
+
+  // ambient declarations must not have the `declare` keyword when in
+  // a namespace
+  sourceFile.forEachChild(node => {
+    if (TypeGuards.isAmbientableNode(node)) {
+      node.setHasDeclareKeyword(false);
+    }
+  });
+  const namespaceDeclaration = destFile.addNamespace({
+    name: `"${namespace}"`,
+    hasDeclareKeyword: true,
+    declarationKind: NamespaceDeclarationKind.Module
+  });
+  // for some reason, types and interfaces that are "declare"d are not parsed
+  // as ambientable nodes, therefore we have to strip the declares off them
+  // manually.
```

Comment:
They should be (see [here](https://github.com/dsherret/ts-simple-ast/blob/f3dda61346587a4a96c7c9a7e515ff071f167053/src/utils/TypeGuards.ts#L72)). Perhaps it should go into the children when it's a `NamesapceDeclaration` above on line 84 (or use `forEachDescendant`)?

Reply:
Nevermind. My bad! As discussed, this is fixed in 16.0.3.